src/opf/dcp.jl

```diff
@@ -1,263 +1,171 @@
-    L = length(ref[:load])
-    bus_loads = [
-        [ref[:load][l] for l in ref[:bus_loads][i]]
-        for i in 1:N
-    ]
-    bus_shunts = [
-        [ref[:shunt][s] for s in ref[:bus_shunts][i]]
-        for i in 1:N
-    ]
+    N, E, G = data.N, data.E, data.G
+    gs = data.gs
+    pd = data.pd
+    i0 = data.ref_bus
+    bus_arcs_fr, bus_arcs_to = data.bus_arcs_fr, data.bus_arcs_to
+    bus_gens = data.bus_gens
+    pgmin, pgmax = data.pgmin, data.pgmax
+    c0, c1, c2 = data.c0, data.c1, data.c2
+    gen_status = data.gen_status
+    bus_fr, bus_to, b = data.bus_fr, data.bus_to, data.b
+    dvamin, dvamax, smax = data.dvamin, data.dvamax, data.smax
+    branch_status = data.branch_status
 
     model = JuMP.GenericModel{T}(optimizer)
-    model.ext[:opf_model] = DCOPF  # for internal checks
+    model.ext[:opf_model] = DCOPF
 
     #
     #   I. Variables
     #
+    
+    # nodal voltage
+    @variable(model, va[1:N])
 
-    # bus voltage angle
-    JuMP.@variable(model, va[1:N])
-
-    # generator active dispatch
-    JuMP.@variable(model, ref[:gen][g]["pmin"] <= pg[g in 1:G] <= ref[:gen][g]["pmax"])
-
-    # branch flows
-    JuMP.@variable(model, -ref[:branch][l]["rate_a"] <= pf[(l,i,j) in ref[:arcs_from]] <= ref[:branch][l]["rate_a"])
-
-    # https://github.com/lanl-ansi/PowerModels.jl/blob/d9c00d228a4e875ac3426425ad6c8f8338309efa/src/form/dcp.jl#L255-L258
-    # pf_expr = Dict{Any, Any}( ((l,i,j), pf[(l,i,j)]) for (l,i,j) in ref[:arcs_from] )
-    # pf_expr = merge(pf_expr, Dict( ((l,j,i), -1.0*pf[(l,i,j)]) for (l,i,j) in ref[:arcs_from] ))
-    pf_expr = [((l,i,j), pf[(l,i,j)]) for (l,i,j) in ref[:arcs_from]]
-    pf_expr = vcat(pf_expr, [((l,j,i), -1.0*pf[(l,i,j)]) for (l,i,j) in ref[:arcs_from]])
+    # Active and reactive dispatch
+    @variable(model, pg[g in 1:G])
 
-    model[:pf] = pf = JuMP.Containers.DenseAxisArray(collect(getfield.(pf_expr, 2)), collect(getfield.(pf_expr, 1)))
+    # Branch flows
+    @variable(model, pf[e in 1:E])
 
-    #
+    # 
     #   II. Constraints
     #
 
+    # Generation bounds (both zero if generator is off)
+    set_lower_bound.(pg, gen_status .* pgmin)
+    set_upper_bound.(pg, gen_status .* pgmax)
+
+    # Flow bounds (both zero if branch is off)
+    set_lower_bound.(pf, branch_status .* -smax)
+    set_upper_bound.(pf, branch_status .* smax)
+
     # Slack bus
-    JuMP.@constraint(model, slack_bus, va[i0] == 0.0)
+    @constraint(model, slack_bus, va[i0] == 0.0)
 
     # Nodal power balance
-    JuMP.@constraint(model,
+    @constraint(model,
         kirchhoff[i in 1:N],
-        sum(pg[g] for g in ref[:bus_gens][i])
-        - sum(pf[a] for a in ref[:bus_arcs][i]) 
-        ==
-        sum(load["pd"] for load in bus_loads[i])
-        + sum(shunt["gs"] for shunt in bus_shunts[i])
+        sum(pg[g] for g in bus_gens[i] if gen_status[g])
+        - sum(pf[a] for a in bus_arcs_fr[i] if branch_status[a])
+        - sum(-pf[a] for a in bus_arcs_to[i] if branch_status[a])
```

Comment:
I put it this way to be consistent with the ACOPF and SOCOPF power balance. I suppose it would be cleaner to use `@expression(model, pt[e in 1:E], -pf[e])` here. See also my comment below about multiplying by `branch_status[e]` instead of filtering.

Reply:
I agree with consistence in the context of comparing AC vs SOC vs DC.
There is however value in having a simple formulation that's easier to read and process. I think the double minus sign increases the risk of mis-reading.